### PR TITLE
Php 8.0 support

### PR DESCRIPTION
**Pull request by hugochinchilla**
Hi, I started working on having php 8.0 compatibility:

 * ~~replaced `dflydev/fig-cookies` with `yiisoft/cookies`~~
 * ~~dropped `php7.3` because of `yiisoft/cookies`~~
 * updated `dflydev/fig-cookies` to 3.0
 * depend on `doctrine/mongodb-odm` instead of `doctrine/mongodb-odm-bundle`
 * fix `doctrine/mongodb-odm` on a specific commit in master as there is no release with `php8.0` support
 * `sentry/sentry-symfony` updated to 4.0, which causes sentry to update to 3.x, some changes to SentryMiddleware where needed

I have tested everything with php 7.4 and 8.0

I think to release this a major version change would be required because of the changes in Sentry and dropping php 7.3

### CODE REVIEW

**Comment by hugochinchilla** [comment]
`dflydev/fig-cookies` released today version 3.0 with php 8.0 support, I have reverted the changes to stop using that library, also I have brought back support for php 7.3.

Still a major version should be considered because of the implications of updating sentry to 3.x

**Comment by hugochinchilla** [comment]
As per https://github.com/doctrine/mongodb-odm/issues/2273 we can expect a `doctrine/mongodb-odm` release today or tomorrow

**Comment by Baldinof** [comment]
Hi, thanks a lot for your work :)

I will try to find some time to review it by the end of this weekend!

**Comment by Baldinof** [comment]
When I wrote the Sentry Middleware, it was not possible to passes an arbitratry PSR request to the Sentry integration, it was always created from superglobals.

Now it seems it can be configured with a `RequestFetcher`, it's implemented in the bundle here : https://github.com/getsentry/sentry-symfony/blob/master/src/Integration/RequestFetcher.php

With RoadRunner we already have a PSR request, it's a waste of time to recreate one from the Symfony one.

What do you think if we:

- add a `RequestFetcher` that decorate the one from Sentry Bundle (the decorator should run the default fetcher if the current runtime is not RoadRunner, see [StreamedResponseListener](https://github.com/Baldinof/roadrunner-bundle/blob/c377846b5fe7529b207a4b803c91d752a532aa2f/src/EventListener/StreamedResponseListener.php#L31))
- rewrite the Sentry Middleware to simply configure the PSR request on the decorated `RequestFetcher` before handling the request (and flush the client)

I don't know if this should be a major release or if a `conflict` section in composer.json can be enough.

**Comment by hugochinchilla** [comment]
I think this bundle should remain independent from sentry bundle, not everyone using this bundle will use the sentry bundle, they will have it installed because it is a dependendy but this doesn't mean they will have it enabled (in the same way I have doctrine/mongodb-odm-bundle installed but not enabled).

So I think it's better to have independende and recreathe the PSR7 request, it's not that much time, completly neglibible if compared with the real causes of latency in any real web-app, like database queries or any other IO operation. And has the benefit to keeping all core features local to this repository which is easier to test and easier to understand for new people wanting to contribute.

A conflict in composer.json would be enought, I didn't think of that.

**Comment by Baldinof** [comment]
I was not proposing to make sentry mandatory here, just clearing a big part of the `SentryMiddleware` because it seems that we can rely on the code on their side.

Howeover it's not a big deal, for the php 8 support I can merge this with a `conflict` section in `composer.json`.

File: tests/BaldinofRoadRunnerBundleTest.php [hunk 1]
```diff
@@ -17,7 +17,7 @@
 
 class BaldinofRoadRunnerBundleTest extends TestCase
 {
-    public function test_it_loads_sentry_middleware_if_needed()
+    public function testItLoadsSentryMiddlewareIfNeeded()
```
Comment (Baldinof) [inline tests/BaldinofRoadRunnerBundleTest.php:20]: Can you revert all renamed tests and add `'php_unit_method_casing' => false` in `.php_cs.dist`

Comment (hugochinchilla) [reply tests/BaldinofRoadRunnerBundleTest.php]: Done!

File: composer.json [hunk 2]
```diff
@@ -44,10 +44,10 @@
         "phpspec/prophecy-phpunit": "^2.0",
         "friendsofphp/php-cs-fixer": "^2.16",
         "phpstan/phpstan": "^0.12.2",
-        "sentry/sentry-symfony": "^3.4",
+        "sentry/sentry-symfony": "^4.0",
         "symfony/framework-bundle": "^4.0||^5.0",
         "nyholm/psr7": "^1.2",
-        "doctrine/mongodb-odm-bundle": "^4.1",
+        "doctrine/mongodb-odm": "dev-master#fa967eadf9e226923a97b145e8018f94d6580f09",
```
Comment (Baldinof) [inline composer.json:50]: 2.2 seems to be released with PHP8 support.

Also can you add the `conflict` section?

Comment (hugochinchilla) [reply composer.json:50]: Ouch, this was alreadi fixed but I force-pushed :facepalm: 

I just realized I don't know how to do the conflict section, if we are targeting `sentry/sentry-symfony >= 4.0` which already targets `sentry/sentry => 3.0` I think it's impossible for this package to be installed with sentry 2.X. The problem is that upgrading the sentry-bundle and sentry will require manual intervention. Anyway, if they already changed to a major version, why should we do it? The users will have to edit their composer.json to allow upgrading it and thus they will expect things to require some manual changes.

Comment (Baldinof) [reply composer.json]: This package does not depend directy on `sentry/sentry-symfony` it's only in `require-dev` to test the integration.

You can already install `4.0` but the integration will not work find without your changes.

Maybe we can go for another solution and make the code compatible with sentry 2 & 3.

Maybe a static utility class can wraps `flush` & `setIpAdress` to do the right thing depending on the currently installed version. What do you think?

Comment (hugochinchilla) [reply composer.json]: ok, I did it. The most problematic part was providing different mocks for the interface `TransportFactoryInterface` in the `SentryMiddlewareTest`. Also I had to add some ignore rules for phpstan as it complains from the use of non-existing methods.

Comment (hugochinchilla) [reply composer.json]: phpstan will work fine if sentry 3.x is used, for sentry 2.x some rules should be added to `ignoreErrors`

File: src/EventListener/SentryListener.php [hunk 3]
```diff
@@ -21,8 +21,8 @@ public function onWorkerStop(WorkerStopEvent $event): void
     {
         $client = $this->hub->getClient();
 
-        if ($client instanceof FlushableClientInterface) {
-            $client->flush();
+        if ($client instanceof ClientInterface) {
```
Comment (Baldinof) [inline src/EventListener/SentryListener.php:24]: You can also use the helper here.

File: src/Helpers/SentryHelper.php [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+use Sentry\FlushableClientInterface;
+use Sentry\UserDataBag;
+
+class SentryHelper
+{
+    public static function isVersion3(): bool
+    {
+        $version = PrettyVersions::getVersion('sentry/sentry')->getPrettyVersion();
+
+        return version_compare($version, '3.0.0', '>=');
```
Comment (Baldinof) [inline src/Helpers/SentryHelper.php:19]: The result of this can be cached in a static property.

File: composer.json [hunk 5]
```diff
@@ -17,7 +17,7 @@
         }
     },
     "require": {
-        "php": ">=7.3",
+        "php": ">=7.3 || >= 8.0",
```
Comment (Baldinof) [inline composer.json:20]: `>= 7.3` already allows PHP 8

File: composer.json (hunk 2 above)
Comment (Baldinof) [reply composer.json]: Cool :)

I sent a last batch of comments and we are good!

It's not a problem for PHPStan.

File: composer.json (hunk 5 above)
Comment (hugochinchilla) [reply composer.json]: ok

File: src/Helpers/SentryHelper.php (hunk 4 above)
Comment (hugochinchilla) [reply src/Helpers/SentryHelper.php]: done!

File: src/EventListener/SentryListener.php (hunk 3 above)
Comment (hugochinchilla) [reply src/EventListener/SentryListener.php]: oh, I missed this one, done!

**Comment by Baldinof** [comment]
Thank you @hugochinchilla :)

**Comment by hugochinchilla** [comment]
Thanks! Can we have a new release soon?

**Comment by Baldinof** [comment]
I just released 1.4.0.

**Pull request merged**